crates/radix-engine-toolkit/src/models/canonical_address_types.rs

```diff
@@ -0,0 +1,342 @@
+                            None
+                        }
+                    } else {
+                        None
+                    }
+                }
+
+                pub fn try_from_bech32(bech32: &str) -> Option<Self> {
+                    let network_id = network_id_from_address_string(bech32)?;
+
+                    let decoder = AddressBech32Decoder::new(
+                        &network_definition_from_network_id(network_id),
+                    );
+                    if let Ok((entity_type, mut full_data)) =
+                        decoder.validate_and_decode(bech32)
+                    {
+                        full_data.remove(0); // skip entity type
+                        if let Ok(node_id) = full_data.as_slice().try_into() {
+                            Self::try_from_node_id(
+                                &NodeId::new(entity_type as u8, node_id),
+                                network_id,
+                            )
+                        } else {
+                            None
+                        }
+                    } else {
+                        None
+                    }
+                }
+
+                fn is_entity_type_valid(entity_type: EntityType) -> bool {
+                    matches!(
+                        entity_type,
+                        $entity_type
+                    )
+                }
+            }
+
+            impl CanonicalAddress for [<Canonical $name Address>] {
+                fn entity_type(&self) -> EntityType {
+                    // Safe to unwrap as entity type is validated during this objcet creation.
+                    self.address.entity_type().unwrap()
+                }
+
+                fn network_id(&self) -> NetworkId {
+                    self.network_id
+                }
+
+                fn to_bech32(&self) -> Result<String, CanonicalAddressError> {
+                    let encode = AddressBech32Encoder::new(
+                        &network_definition_from_network_id(self.network_id),
+                    );
+                    encode
+                        .encode(self.address.as_bytes())
+                        .map_err(|_| CanonicalAddressError::FailedToEncodeBech32)
+                }
+            }
+
+            impl FromStr for [<Canonical $name Address>] {
+                type Err = CanonicalAddressError;
+
+                fn from_str(bech32: &str) -> Result<Self, Self::Err> {
+                    Self::try_from_bech32(bech32)
+                        .ok_or(CanonicalAddressError::FailedToDecodeBech32)
+                }
+            }
+
+            impl Display for [<Canonical $name Address>] {
+                fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                    formatter.write_str(&self.to_bech32().map_err(|_| fmt::Error)?)
+                }
+            }
+        }
+    };
+
+    ($name: ident, $entity_type: pat, $fungibility_validator: expr) => {
```

Comment:
I would prefer if instead of this being an additional arm in the macro we turn it into an implementation we write by hand as this isn't a general case, just a special case for resource address.